### PR TITLE
Add new device: IKEA SOMRIG E2213

### PR DESCRIPTION
**Pull request by thatslolo**
I was able to snatch a few of the yet (to my knowledge) unreleased SOMRIG Shortcut buttons and got it successfully configured.
The button actions are transferred by `heimanSpecificScenes`. Not sure how the handle them for now.
Advice how to use them would be greatly appreciated.

### CODE REVIEW

**Comment by Koenkk** [comment]
> The button actions are transferred by heimanSpecificScenes. Not sure how the handle them for now.


Can you provide the debug logging when these actions are sent?

See https://www.zigbee2mqtt.io/guide/usage/debug.html on how to enable debug logging.

**Comment by Koenkk** [comment]
Could you also provide the herdsman debug logging of this? One click is enough.

See https://www.zigbee2mqtt.io/guide/usage/debug.html on how to enable the herdsman debug logging. Note that this is only logged to STDOUT and not to log files.

**Comment by thatslolo** [comment]
when 1 pressed short:

`2023-11-29T09:12:48.336Z zigbee-herdsman:adapter:zStack:unpi:parser <-- [254,26,68,129,0,0,128,252,225,233,1,1,0,120,0,192,133,183,0,0,6,21,124,17,36,1,0,67,18,28,55]
2023-11-29T09:12:48.336Z zigbee-herdsman:adapter:zStack:unpi:parser --- parseNext [254,26,68,129,0,0,128,252,225,233,1,1,0,120,0,192,133,183,0,0,6,21,124,17,36,1,0,67,18,28,55]
2023-11-29T09:12:48.336Z zigbee-herdsman:adapter:zStack:unpi:parser --> parsed 26 - 2 - 4 - 129 - [0,0,128,252,225,233,1,1,0,120,0,192,133,183,0,0,6,21,124,17,36,1,0,67,18,28] - 55
2023-11-29T09:12:48.337Z zigbee-herdsman:adapter:zStack:znp:AREQ <-- AF - incomingMsg - {"groupid":0,"clusterid":64640,"srcaddr":59873,"srcendpoint":1,"dstendpoint":1,"wasbroadcast":0,"linkquality":120,"securityuse":0,"timestamp":12027328,"transseqnumber":0,"len":6,"data":{"type":"Buffer","data":[21,124,17,36,1,0]}}
2023-11-29T09:12:48.340Z zigbee-herdsman:controller:log Received 'raw' data '{"clusterID":64640,"data":{"type":"Buffer","data":[21,124,17,36,1,0]},"address":59873,"endpoint":1,"linkquality":120,"groupID":0,"wasBroadcast":false,"destinationEndpoint":1}'
Zigbee2MQTT:debug 2023-11-29 10:12:48: Received Zigbee message from '0x5cc7c1fffeb1e219', type 'raw', cluster 'heimanSpecificScenes', data '{"data":[21,124,17,36,1,0],"type":"Buffer"}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-29 10:12:48: No converter available for 'E2213' with cluster 'heimanSpecificScenes' and type 'raw' and data '{"data":[21,124,17,36,1,0],"type":"Buffer"}'
2023-11-29T09:12:48.348Z zigbee-herdsman:adapter:zStack:unpi:parser --- parseNext []
2023-11-29T09:12:49.144Z zigbee-herdsman:adapter:zStack:unpi:parser <-- [254,26,68,129,0,0,128,252,225,233,1,1,0,127,0,219,74,184,0,0,6,21,124,17,37,3,0,67,18,28,232]
2023-11-29T09:12:49.145Z zigbee-herdsman:adapter:zStack:unpi:parser --- parseNext [254,26,68,129,0,0,128,252,225,233,1,1,0,127,0,219,74,184,0,0,6,21,124,17,37,3,0,67,18,28,232]
2023-11-29T09:12:49.145Z zigbee-herdsman:adapter:zStack:unpi:parser --> parsed 26 - 2 - 4 - 129 - [0,0,128,252,225,233,1,1,0,127,0,219,74,184,0,0,6,21,124,17,37,3,0,67,18,28] - 232
2023-11-29T09:12:49.146Z zigbee-herdsman:adapter:zStack:znp:AREQ <-- AF - incomingMsg - {"groupid":0,"clusterid":64640,"srcaddr":59873,"srcendpoint":1,"dstendpoint":1,"wasbroadcast":0,"linkquality":127,"securityuse":0,"timestamp":12077787,"transseqnumber":0,"len":6,"data":{"type":"Buffer","data":[21,124,17,37,3,0]}}
2023-11-29T09:12:49.148Z zigbee-herdsman:controller:log Received 'raw' data '{"clusterID":64640,"data":{"type":"Buffer","data":[21,124,17,37,3,0]},"address":59873,"endpoint":1,"linkquality":127,"groupID":0,"wasBroadcast":false,"destinationEndpoint":1}'
Zigbee2MQTT:debug 2023-11-29 10:12:49: Received Zigbee message from '0x5cc7c1fffeb1e219', type 'raw', cluster 'heimanSpecificScenes', data '{"data":[21,124,17,37,3,0],"type":"Buffer"}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-29 10:12:49: No converter available for 'E2213' with cluster 'heimanSpecificScenes' and type 'raw' and data '{"data":[21,124,17,37,3,0],"type":"Buffer"}'
2023-11-29T09:12:49.156Z zigbee-herdsman:adapter:zStack:unpi:parser --- parseNext []`

**Comment by Koenkk** [comment]
Thanks, I've added the cluster now, with this you will now get the parsed messages instead of `raw` which you can use to write the `fromZigbee` converters.

Changes will be available in the dev branch in a few hours from now. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Comment by thatslolo** [comment]
After some digging I found this issue: Koenkk/zigbee2mqtt#16808 which discussed the implementation of the dot buttons of the symfonsik controller lengthily. 
The current implementation by @ilueckel from #5583 would work already with minimal changes without the implementation of the new `tradfriButton` cluster.
```
        {
            cluster: 'heimanSpecificScenes',
            type: 'raw',
            convert: (model, msg, publish, options, meta) => {
                let button;
                let action;
                switch (msg.endpoint.ID) {
                case 1: button = '1'; break; // 1 dot
                case 2: button = '2'; break; // 2 dot
                }
                switch (msg.data[4]) {
                case 1: action = 'initial_press'; break;
                case 2: action = 'long_press'; break;
                case 3: action = 'short_release'; break;
                case 4: action = 'long_release'; break;
                case 6: action = 'double_press'; break;
                }
    
                return {action: `dots_${button}_${action}`};
            },
        }
```

Besides, the new cluster broke the dot buttons for the Symfonisk remote:

```
Debug 2023-12-01 17:05:06Received Zigbee message from '', type 'commandAction1', cluster 'tradfriButton', data '{"data":0}' from endpoint 3 with groupID 0
Debug 2023-12-01 17:05:06No converter available for 'E2123' with cluster 'tradfriButton' and type 'commandAction1' and data '{"data":0}'
```

**Comment by Koenkk** [comment]
This will not work anymore with z2m 1.34.0, please test again that version.

**Comment by Koenkk** [comment]
Can you try with this converter: https://gist.github.com/Koenkk/ad301d469171bfe1b977a4e021e821cc#file-ext-js-L33 ?

**Comment by thatslolo** [comment]
The converter throws the exception for the symfonisk as described in https://github.com/Koenkk/zigbee2mqtt/discussions/20003

**Comment by Koenkk** [comment]
Pushed a fix for this, please check again with latest dev

Changes will be available in the [dev branch](https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html) in a few hours from now.

**Comment by MattWestb** [comment]
Looks like this device is using Matter ZCL like Symfonisk gen 2 on manufacture cluster (Likely IKEA is updating the firmware using matter cluster in near future on both devices then HA can taking care of the commands) with production firmware (not factory) and only need EP 1 and 2 being bonded then looking in sniffs from it so shall being easy getting initial, single double long and long release working.

File: src/devices/ikea.ts [hunk 1]
```diff
@@ -281,10 +281,26 @@ const fzLocal = {
     } satisfies Fz.Converter,
     ikea_dots_click_v2: {
         // For remotes with firmware 1.0.32 (20221219)
+        cluster: 'heimanSpecificScenes',
```
[the remark this answers is not in the document]
Comment (thatslolo) [reply src/devices/ikea.ts:284]: Me as well. Must be an error from fetching upstream.

**Comment by thatslolo** [comment]
should be ready now

File: src/devices/ikea.ts [hunk 2]
```diff
@@ -296,6 +296,22 @@ const fzLocal = {
...
+            const button = utils.getFromLookup(msg.endpoint.ID, {1: '1', 2: '2'});
+            const lookup = {
+                commandAction1: 'initial_press',
+                commandAction2: 'long_press',
+                commandAction3: 'short_release',
+                commandAction4: 'long_release',
+                commandAction6: 'double_press',
+            };
+            const action = utils.getFromLookup(msg.type, lookup);
+            return {action: `dots_${button}_${action}`};
```
Comment (Koenkk) [inline src/devices/ikea.ts:312]: I think we can remove the `dots_` prefix here (since it's always there). After that this can be merged.

Comment (thatslolo) [reply src/devices/ikea.ts]: Good point!

**Comment by Koenkk** [comment]
Thanks!

**Comment by kafisc1** [comment]
Is there a working externernal converter at the moment so I can bridge the time until there is new Home Assistant addon edge release? Thank you in advance

**Comment by b2un0** [comment]
@kafisc1 I created the following converter on this PR.

on button press, I got the following error

## single button 1 press
```
Exception while calling fromZigbee converter: Expected one of: commandAction1, commandAction2, commandAction3, commandAction4, commandAction6, got: 'commandAction1'}
Exception while calling fromZigbee converter: Expected one of: commandAction1, commandAction2, commandAction3, commandAction4, commandAction6, got: 'commandAction3'}
```

in https://github.com/Koenkk/zigbee2mqtt/discussions/20003 the problem was solved by a not referenced commit to `dev`- I can't find them.

@Koenkk can you tell uns where we can fix this issue?


```js
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const utils = require('zigbee-herdsman-converters/lib/utils');
const ota = require('zigbee-herdsman-converters/lib/ota');
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    ikea_dots_click_v1_somrig: {
        // For remotes with firmware 1.0.012 (20211214)
        cluster: 64639,
        type: 'raw',
        convert: (model, msg, publish, options, meta) => {
            if (!Buffer.isBuffer(msg.data)) return;
            let action;
            const button = msg.data[5];
            switch (msg.data[6]) {
                case 1:
                    action = 'initial_press';
                    break;
                case 2:
                    action = 'double_press';
                    break;
                case 3:
                    action = 'long_press';
                    break;
            }

            return {action: `dots_${button}_${action}`};
        },
    },
    ikea_dots_click_v2_somrig: {
        cluster: 'tradfriButton',
        type: ['commandAction1', 'commandAction2', 'commandAction3', 'commandAction4', 'commandAction6'],
        convert: (model, msg, publish, options, meta) => {
            const button = utils.getFromLookup(msg.endpoint.ID, {1: '1', 2: '2'});
            const lookup = {
                commandAction1: 'initial_press',
                commandAction2: 'long_press',
                commandAction3: 'short_release',
                commandAction4: 'long_release',
                commandAction6: 'double_press',
            };
            const action = utils.getFromLookup(msg.type, lookup);
            return {action: `${button}_${action}`};
        }
    },
}

const definition = {
    zigbeeModel: ['SOMRIG shortcut button'],
    model: 'E2213',
    vendor: 'IKEA',
    description: 'SOMRIG shortcut button',
    fromZigbee: [fz.battery, fzLocal.ikea_dots_click_v1_somrig, fzLocal.ikea_dots_click_v2_somrig],
    toZigbee: [tz.battery_percentage_remaining],
    ota: ota.tradfri,
    exposes: [
        e.battery().withAccess(ea.STATE_GET), e.action(['dots_1_initial_press',
            'dots_2_initial_press', 'dots_1_long_press', 'dots_2_long_press',
            'dots_1_short_release', 'dots_2_short_release', 'dots_1_long_release']),
    ],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint1 = device.getEndpoint(1);
        const endpoint2 = device.getEndpoint(2);
        await reporting.bind(endpoint1, coordinatorEndpoint, ['tradfriButton', 'genPollCtrl']);
        await reporting.bind(endpoint2, coordinatorEndpoint, ['tradfriButton']);
        await reporting.batteryVoltage(endpoint1);
    },
};

module.exports = definition;


```

**Comment by b2un0** [comment]
@kafisc1 

here you go

```js
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const utils = require('zigbee-herdsman-converters/lib/utils');
const ota = require('zigbee-herdsman-converters/lib/ota');
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const e = exposes.presets;
const ea = exposes.access;

const fzLocal = {
    ikea_dots_click_v2_somrig: {
        cluster: 'tradfriButton',
        type: ['commandAction1', 'commandAction2', 'commandAction3', 'commandAction4', 'commandAction6'],
        convert: (model, msg, publish, options, meta) => {
            const button = utils.getFromLookup(msg.endpoint.ID, {1: '1', 2: '2'});
            const lookup = {
                commandAction1: 'initial_press',
                commandAction2: 'long_press',
                commandAction3: 'short_release',
                commandAction4: 'long_release',
                commandAction6: 'double_press',
            };
            const action = lookup[msg.type] ?? lookup[msg.type.toLowerCase()] ?? lookup[msg.type.toUpperCase()];
            return {action: `${button}_${action}`};
        }
    },
}

const definition = {
    zigbeeModel: ['SOMRIG shortcut button'],
    model: 'E2213',
    vendor: 'IKEA',
    description: 'SOMRIG shortcut button',
    fromZigbee: [fz.battery, fzLocal.ikea_dots_click_v2_somrig],
    toZigbee: [tz.battery_percentage_remaining],
    ota: ota.tradfri,
    exposes: [
        e.battery().withAccess(ea.STATE_GET), e.action(['dots_1_initial_press',
            'dots_2_initial_press', 'dots_1_long_press', 'dots_2_long_press',
            'dots_1_short_release', 'dots_2_short_release', 'dots_1_long_release']),
    ],
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint1 = device.getEndpoint(1);
        const endpoint2 = device.getEndpoint(2);
        await reporting.bind(endpoint1, coordinatorEndpoint, ['tradfriButton', 'genPollCtrl']);
        await reporting.bind(endpoint2, coordinatorEndpoint, ['tradfriButton']);
        await reporting.batteryVoltage(endpoint1);
    },
};

module.exports = definition;

```

**Comment by b2un0** [comment]
@kafisc1 ca be found here
https://github.com/b2un0/z2m-device-converter/

**Comment by jcarelli** [comment]
Hi, Hopefully you can help?  Using HS$ & ZIGBEE Plus, and your external controller. .   
I'm getting the follow type of warning on the various button presses. "Cannot match value 2_short_release to an enum value".
The devices pairs and reports back the button presses. any help would be appreciated.  Thanks, 

sorry if I'm in the wrong place,

**Pull request merged**